### ISSUE TITLE
browser/jellyfish.umd.js to be exported as global mode

### ISSUE THREAD

**Issue by fuxingloh** [DeFiCh/jellyfish#52]
> Moved out of scope from #17 due to waiting for upstream fixes and implementation.

#### What would you like to be added:

```ts
import jellyfish from 'jellyfish.ts'
window.jellyfish = jellyfish
```

#### Why is this needed:

To allow browser users to do this:

```html
<script src="https://unpkg.com/@defichain/jellyfish@latest/dist/jellyfish.umd.js"/>
<script>
  var client = new jellyfish.Client('http://ocean.defichain.com')
  client.mining.getMintingInfo().then(info => {
    console.log(info)
  })
</script>
```

**Comment by fuxingloh** [comment]
When completed: 
- [ ] Update docs website
- [ ] Update `README.md`

**Comment by fuxingloh** [comment]
Still waiting for support from parcel: https://v2.parceljs.org/getting-started/migration/#--global

**Comment by IhorSavva** [comment]
Hey @fuxingloh! 
Are there any updates? 
I really look forward to using the library in a browser. If now there is some opportunity to do this, please tell me :)

**Comment by fuxingloh** [comment]
> Hey @fuxingloh!
> Are there any updates?
> I really look forward to using the library in a browser. If now there is some opportunity to do this, please tell me :)

Hey @IhorSavva 

The `jellyfish.umd.js` output bundles together `jellyfish-api-core` & `jellyfish-api-jsonrpc` which are only 2/15 of the total modules available. They are created merely for the browser for interacting with the API, not very useful other than doing just that. Still waiting for upstream library to support it as switching to another bundler (e.g. rollup) is quite a pain.

https://github.com/DeFiCh/jellyfish/blob/015ec69bf41c73376ff1f659645e0095aceb7a42/packages/jellyfish/package.json#L37-L45

To use jellyfish packages, I suggest you `import` what you need via the esm exports. E.g. if you need the wallet component, you import `import { generateMnemonic } from '@defichain/jellyfish-wallet-mnemonic'`. If your target is a website then you can use `webpack` to do it.

We do that for our wallet and explorer project: [DeFiCh/wallet package.json](https://github.com/DeFiCh/wallet/blob/d07e274f27cbf2f32757454c6e54ba461273efe8/package.json#L20-L26)

Some examples for reference:

1. Vue Website: https://github.com/fuxingloh/defi-txn-composer
2. React Website: https://github.com/DeFiCh/explorer
3. React Native: https://github.com/DeFiCh/wallet

**Comment by IhorSavva** [comment]
@fuxingloh, thanks for the answer!
See, I need the createRawTx and signRawTx functionality.
Tell me, can I somehow use this functionality in the browser (html + js)? Or, at least just for example in VueJS
And another question. Do you execute create and sing on the client, or is it the same request to the RPC itself?

**Comment by fuxingloh** [comment]
> @fuxingloh, thanks for the answer!
> See, I need the createRawTx and signRawTx functionality.
> Tell me, can I somehow use this functionality in the browser (html + js)? Or, at least just for example in VueJS
> And another question. Do you execute create and sing on the client, or is it the same request to the RPC itself?

Yup, you can, this is a full end-to-end test with creating and signing a raw transaction.
https://github.com/DeFiCh/jellyfish/blob/main/packages/jellyfish-transaction-builder/__tests__/raw_crafting.ts

Yes, it will work on any platform (batteries not included), it's ubiquitous and written platform-agnostic with fully-featured implementation. (key signing only support Bech32 only for now)

It's non-custodial, so the keys for creating and signing a transaction do not have to be on the node. So no it's not on the RPC, this library does everything manually from scratch, the RPC is just a small part of what jellyfish do.
E.g. this is a classic single key wallet https://github.com/DeFiCh/jellyfish/blob/main/packages/jellyfish-wallet-classic/__tests__/wallet_classic.test.ts

**Comment by IhorSavva** [comment]
@fuxingloh Thanks for your answers, it helps a lot!
Tell me how can I use testnet when creating a transaction?
And another question, does your library already know how to work with hd-wallets? (If so, could you show me how I can work with this?)

**Comment by fuxingloh** [comment]
For creating a transaction, the tests cases are written exactly as such to show you how to create a transaction:
https://github.com/DeFiCh/jellyfish/blob/main/packages/jellyfish-transaction-builder/__tests__/raw_crafting.ts

Yes, the `hd-wallet` code is here:
https://github.com/DeFiCh/jellyfish/blob/main/packages/jellyfish-wallet/__tests__/wallet_hd_node.test.ts#L4-L9

**Comment by IhorSavva** [comment]
@fuxingloh About tx creation. I mean: do you library support creating a transaction in the testnet network? If so, how can I do this?
How can I create a transaction without knowing the recipient's private key?(how I can get public_key having only an address)
https://github.com/DeFiCh/jellyfish/blob/559eab98e9f9270a1d0f167942e5c4f167d9141a/packages/jellyfish-transaction-builder/__tests__/raw_crafting.ts#L40-L43

https://github.com/DeFiCh/jellyfish/blob/559eab98e9f9270a1d0f167942e5c4f167d9141a/packages/jellyfish-transaction-builder/__tests__/raw_crafting.ts#L65-L79

**Comment by fuxingloh** [comment]
Sorry, it's really difficult to answer some of your questions. Transactions are [network](https://github.com/DeFiCh/jellyfish/blob/main/packages/jellyfish-network) agnostic. So are addresses, publicKey is address agnostic.

**Comment by IhorSavva** [comment]
Hey, @fuxingloh! 
Okay, let me rephrase.
Could you show me some sample code so that I can create and sign a transaction knowing:
1.about the sender (utxo and private keys)
2.about the recipient (address, amount)

**Comment by fuxingloh** [comment]
> Hey, @fuxingloh!
> Okay, let me rephrase.
> Could you show me some sample code so that I can create and sign a transaction knowing:
> 1.about the sender (utxo and private keys)
> 2.about the recipient (address, amount)

both 1 & 2

https://github.com/DeFiCh/jellyfish/blob/ad6e42290aca0719ee4f27cb7e9c212bcd9a2a2f/packages/jellyfish-transaction-builder/__tests__/raw_crafting.test.ts#L78-L102

**Comment by IhorSavva** [comment]
@fuxingloh, but in order to create a transaction, you need to know the public key for output (this can be seen from the attached link). However, how can we get this from just an address?
https://github.com/DeFiCh/jellyfish/blob/559eab98e9f9270a1d0f167942e5c4f167d9141a/packages/jellyfish-transaction-builder/__tests__/raw_crafting.ts#L72

Or am I misunderstanding something?

**Comment by fuxingloh** [comment]
Please open a discussion thread instead, this issue is created purely to track the implementation of global mode.

**Comment by fuxingloh** [comment]
/close as we are not doing UMD anymore, it's hard to maintain long term. 

That's the secondary reason, the primary reason is a full node should never be exposed to the web browser via the RPC protocol. Pretty dangerous to allow your web browser (untrusted runtime) to execute wallet RPC. We will focus on producing ts,esm,cjs builds for the foreseeable future.
